scheme/sevsnp/common.go

```diff
@@ -0,0 +1,199 @@
+	"errors"
+	"fmt"
+	"reflect"
+	"strconv"
+
+	"github.com/google/go-sev-guest/abi"
+	"github.com/google/go-sev-guest/kds"
+	"github.com/google/go-sev-guest/proto/sevsnp"
+	"github.com/veraison/cmw"
+	"github.com/veraison/corim/comid"
+	"github.com/veraison/corim/corim"
+	"github.com/veraison/ratsd/tokens"
+	"github.com/veraison/services/proto"
+)
+
+// measurementByUintKey looks up comid.Measurement in a CoMID by its MKey.
+//
+//	If no measurements are found, returns nil and no error. Otherwise,
+//	returns the error encountered.
+func measurementByUintKey(refVal comid.ValueTriple,
+	key uint64) (*comid.Measurement, error) {
+	for _, m := range refVal.Measurements.Values {
+		if m.Key == nil || !m.Key.IsSet() ||
+			m.Key.Type() != comid.UintType {
+			continue
+		}
+
+		k, err := m.Key.GetKeyUint()
+		if err != nil {
+			return nil, err
+		}
+
+		if k == key {
+			return &m, nil
+		}
+	}
+
+	return nil, nil
+}
+
+// comidFromJson accepts a CoRIM in JSON format and returns a single CoMID
+//
+//	Returns error if there are more than a single CoMID, or passes on
+//	error from corim routine.
+func comidFromJson(buf []byte) (*comid.Comid, error) {
+	extractedCorim, err := corim.UnmarshalUnsignedCorimFromJSON(buf)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(extractedCorim.Tags) > 1 {
+		return nil, errors.New("too many tags")
+	}
+
+	extractedComid, err := corim.UnmarshalComidFromCBOR(
+		extractedCorim.Tags[0],
+		extractedCorim.Profile,
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return extractedComid, nil
+}
+
+func parseCertificateChainFromEvidence(tsm *tokens.TSMReport) (*sevsnp.CertificateChain, error) {
+	var certTable abi.CertTable
+
+	if err := certTable.Unmarshal(tsm.AuxBlob); err != nil {
+		return nil, err
+	}
+
+	return certTable.Proto(), nil
+}
+
+func readCert(cert []byte) ([]byte, error) {
+	if len(cert) == 0 {
+		return nil, errors.New("empty certificate")
+	}
+
+	block, _ := pem.Decode(cert)
+	if block == nil || block.Type != "CERTIFICATE" {
+		return nil, fmt.Errorf("failed to read certificate")
+	}
+	return block.Bytes, nil
+}
+
+func parseAttestationToken(token *proto.AttestationToken) (*tokens.TSMReport, error) {
+	var (
+		err error
+		tsm = new(tokens.TSMReport)
+	)
+
+	switch token.MediaType {
+	case EvidenceMediaTypeTSMCbor:
+		err = tsm.FromCBOR(token.Data)
+		if err != nil {
+			return nil, err
+		}
+	case EvidenceMediaTypeTSMJson:
+		err = tsm.FromJSON(token.Data)
+		if err != nil {
+			return nil, err
+		}
+	case EvidenceMediaTypeRATSd:
+		eat := make(map[string]interface{})
+
+		err = json.Unmarshal(token.Data, &eat)
+		if err != nil {
+			return nil, err
+		}
+
+		cmwBase64, ok := eat["cmw"].(string)
+		if !ok {
+			return nil, fmt.Errorf("cmw not found in eat")
+		}
+
+		cmwJson, err := base64.StdEncoding.DecodeString(cmwBase64)
+		if err != nil {
+			return nil, err
+		}
+
+		cmwCollection, err := cmw.NewCollection("tag:github.com,2025:veraison/ratsd/cmw")
+		if err != nil {
+			return nil, err
+		}
```

Comment:
This is the decoding path, so I am not sure why we need to allocate the collection upfront rather than just declaring a variable of type cmw and letting Unmarshal handle allocation.